hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/transport/server/ratis/ContainerStateMachine.java

```diff
@@ -554,12 +556,12 @@ private ByteString getCachedStateMachineData(Long logIndex, long term,
       }
     } catch (Exception e) {
       metrics.incNumReadStateMachineFails();
-      LOG.error("unable to read stateMachineData:" + e);
+      LOG.error("{} unable to read stateMachineData:", gid, e);
       return completeExceptionally(e);
     }
   }
 
-  private void updateLastApplied() {
+  private synchronized void updateLastApplied() {
```

Comment:
@mukul1987  Thanks for working on this! The patch looks good to me.
I think we can avoid using synchronized by using a concurrent sorted map. Each applyTransaction removes its corresponding entry from the map and we update lastApplied based on the first entry in the map?

Reply:
HDDS-1792 will fix this by using ConcurrentHashSet.